### PR TITLE
Feat/dynamic font loading

### PR DESCRIPTION
**Pull request by fabiowguerra**
Implementing dynamic font loading.
I simply parsed the binary file generated with the font converter and populated the `lv_font_t` struct with the same information the font converter uses.

I created a simple test, that compares a generated C file with a parsed file. I was unsure about where to put the new files, see if it is ok.

### CODE REVIEW

**Comment by puzrin** [comment]
https://docs.lvgl.io/latest/en/html/overview/font.html#add-a-new-font-engine

Could you consider wrap your work not as "lvgl font generator" but as "bin font engine"? That's more universal.

File: src/lv_font/lv_font_loader.c [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+    if(lv_fs_read(fp, &font_header, sizeof(font_header_bin_t), NULL) != LV_FS_RES_OK) {
+        return false;
+    }
+
+    font->base_line = -font_header.descent;
+    font->line_height = font_header.ascent - font_header.descent;
+    font->get_glyph_dsc = lv_font_get_glyph_dsc_fmt_txt;
+    font->get_glyph_bitmap = lv_font_get_bitmap_fmt_txt;
+
+#if !(LVGL_VERSION_MAJOR == 6 && LVGL_VERSION_MINOR == 0)
```
Comment (embeddedt) [inline src/lv_font/lv_font_loader.c]: This code would never run as part of a 6.x system so I think this condition can be removed entirely.

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,8 @@
 - Add `lv_tabview_set_tab_name()` function - used to change a tab's name
 - Add `LV_THEME_MATERIAL_FLAG_NO_TRANSITION` and `LV_THEME_MATERIAL_FLAG_NO_FOCUS` flags
 - Reduce code size by adding: `LV_USE_FONT_COMPRESSED` and `LV_FONT_USE_SUBPX` and applying some optimization
+- Add `lv_font_load()` function - Loads a `lv_font_t` object from a binary font file
+- Add `lv_font_free()` function - Frees the memory allocated by the `lv_font_load()` function
```
Comment (embeddedt) [inline CHANGELOG.md]: We should double check with @kisvegabor, but I'm pretty sure these changelog lines belong under 7.4, not 7.3.

File: src/lv_font/lv_font_loader.c [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+    lv_fs_file_t file;
+    lv_fs_res_t res = lv_fs_open(&file, font_name, LV_FS_MODE_RD);
+
+    if(res == LV_FS_RES_OK) {
+        success = lvgl_load_font(&file, font);
+    }
+
+    if(!success) {
+        LV_LOG_WARN("Error loading font file: %s\n", font_name);
+        lv_font_free(font);
```
Comment (embeddedt) [inline src/lv_font/lv_font_loader.c:110]: Could we add a comment here explaining how this works for a future developer? My understanding from reading the code is that every successfully allocated structure is immediately zeroed - that way any pointers inside can be checked for being non-NULL and then freed.

**Review by embeddedt — CHANGES_REQUESTED** [review]
Looks good in general. I just have a few suggestions (see the review comments).

**Comment by fabiowguerra** [comment]
> https://docs.lvgl.io/latest/en/html/overview/font.html#add-a-new-font-engine
> 
> Could you consider wrap your work not as "lvgl font generator" but as "bin font engine"? That's more universal.

Sorry, I did not understand what you mean about "wrap the work". I did not find the term "font generator" on the code.

What do you think about "bin font parser"?

Fix me if I'm wrong, but I think I did not create a new font engine, since I used the same `get_glyph_dsc` and `get_glyph_bitmap` functions from the current font engine.

I have very little knowledge about fonts, what I did was only parsing a binary file and populating an existing data structure.

File: src/lv_font/lv_font_loader.c (hunk 1 above)
Comment (fabiowguerra) [reply src/lv_font/lv_font_loader.c]: ok i'll change.

File: src/lv_font/lv_font_loader.c (hunk 3 above)
Comment (fabiowguerra) [reply src/lv_font/lv_font_loader.c:110]: Yes. You are right, I comment will be good here.

File: CHANGELOG.md (hunk 2 above)
Comment (fabiowguerra) [reply CHANGELOG.md]: There is a conflict on the CHANGELOG.md, possibly that is the problem. I'll try to resolve.

Comment (fabiowguerra) [reply CHANGELOG.md]: @embeddedt I'm not sure how the branches are used in this project. This PR should goes to the `dev` branch instead of `master`?

Comment (embeddedt) [reply CHANGELOG.md]: That's correct. PRs should be sent to `dev` if they add new features, and to `master` if it's just a simple bugfix.

Comment (kisvegabor) [reply CHANGELOG.md]: > We should double check with @kisvegabor, but I'm pretty sure these changelog lines belong under 7.4, not 7.3.

Yes, I've resolved the conflict and moved these lines to 7.4.

**Comment by puzrin** [comment]
@fabiowguerra 

> Fix me if I'm wrong, but I think I did not create a new font engine, since I used the same `get_glyph_dsc` and `get_glyph_bitmap` functions from the current font engine.
> 
> I have very little knowledge about fonts, what I did was only parsing a binary file and populating an existing data structure.

Yes, this is how your code works now. And i suggest you create `lv_font_t` with "custom" `get_glyph_dsc` and `get_glyph_bitmap` functions funcion. That will use ~ similar decoding, but with this advantages:

- source data can be in flash (as static array)
- no lvgl_format limits
- no memory allocations/free
- no tight coupling with file system
- (?) can be extended later to render from file directly, with low memory footprint (not sure if needed).

Ideally, lvgl was expected [by me] to use bin format by default, but this expects high coverage with tests. And at the moment of convertor workflow design, lvgl did not reached point when tests of some blocks become "mandatory".

At first glance, to convert your code into "font_engine", this 2 things are needed:

- rearrange your code with minimal effort (you already have all working componens)
- probably, add fifo cache for 1-2 last CMAP table results (OR cmap + `get_glyph_dsc()` results).

File: CHANGELOG.md (hunk 2 above)
Comment (kisvegabor) [reply CHANGELOG.md]: I've also fixed the tests, now CI passes.   
d321e0e

**Comment by fabiowguerra** [comment]
@puzrin , thanks for your explanation. I understand that creating a new font engine will be better then my approach.

Maybe it could be added to the LVGL roadmap, reusing part of this code. Unfortunately I have other issues with higher priorities, and can not work on this now.

**Comment by embeddedt** [comment]
I'd say that the current implementation presented now is still better than the previous alternative (no implementation at all). I'd be happy to approve this with a plan to refactor it into a proper engine later on.

**Comment by puzrin** [comment]
@fabiowguerra I understand, no problem

- Your existing commit is certainly "better than nothing"
- It's a good base for migration to "font engine".

The only side effect - it will be breaking change, when replaced with font engine. But it's up to core devs how to organise merge.

File: src/lv_font/lv_font_loader.c (hunk 1 above)
Comment (fabiowguerra) [reply src/lv_font/lv_font_loader.c]: done.

File: src/lv_font/lv_font_loader.c (hunk 3 above)
Comment (fabiowguerra) [reply src/lv_font/lv_font_loader.c:110]: I added some comments here and and a more complete one before the static function `lvgl_load_font`, please see if it's ok.

Comment (embeddedt) [reply src/lv_font/lv_font_loader.c:110]: Looks good.

**Review by embeddedt — APPROVED** [review]

**Comment by kisvegabor** [comment]
Looks good to me as well! Thank you very much!

The only thing is missing is a few lines of description in the [docs](https://docs.lvgl.io/dev/en/html/overview/font.html).

@fabiowguerra do you have capacity to add it?

**Comment by lilingzhiatshenzhen** [comment]
> @fabiowguerra
> 
> > Fix me if I'm wrong, but I think I did not create a new font engine, since I used the same `get_glyph_dsc` and `get_glyph_bitmap` functions from the current font engine.
> > I have very little knowledge about fonts, what I did was only parsing a binary file and populating an existing data structure.
> 
> Yes, this is how your code works now. And i suggest you create `lv_font_t` with "custom" `get_glyph_dsc` and `get_glyph_bitmap` functions funcion. That will use ~ similar decoding, but with this advantages:
> 
> * source data can be in flash (as static array)
> * no lvgl_format limits
> * no memory allocations/free
> * no tight coupling with file system
> * (?) can be extended later to render from file directly, with low memory footprint (not sure if needed).
> 
> Ideally, lvgl was expected [by me] to use bin format by default, but this expects high coverage with tests. And at the moment of convertor workflow design, lvgl did not reached point when tests of some blocks become "mandatory".
> 
> At first glance, to convert your code into "font_engine", this 2 things are needed:
> 
> * rearrange your code with minimal effort (you already have all working componens)
> * probably, add fifo cache for 1-2 last CMAP table results (OR cmap + `get_glyph_dsc()` results).

Is it possible to create the font bin file by C code, I need use C code to transfer and make the font bin file,it is too much to make javascript to C code...thanks

**Pull request merged**